recipe/2024-01-support-private-terraform-repository.md

```diff
@@ -0,0 +1,237 @@
+	insteadOf = https://dev.azure.com
+```
+But it wont work well when we have multiple environments. As global .gitconfig file has the git credentials stored for all the environments and the module repositories may be from different accounts.
+
+Option 2: Saving it as part of the environment resource.
+
+Add a new property `recipeConfig` write-only to store the information about the git credentials, and have a custom action `getRecipeConfiglike` (like `list-secrets`) to get these details. 
+
+```diff
++@secure()
++param username string
++@secure()
++param token string
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'dsrp-resources-env-recipes-context-env'
+  location: 'global'
+  properties: {
+    compute: {
+      ...
+    }
+    providers: {
+      ...
+    }
++   recipeConfig: {
++     terraform: {
++       secrets: {
++         "dev.azure.com": {
++            username : username
++            pat: token
++          }
++       }
++     }
++   }
+    recipes: {      
+      'Applications.Datastores/mongoDatabases':{
+        default: {
+          templateKind: 'terraform'
+          templatePath: 'https://dev.azure.com/test-private-repo'
+        }
+      }
+    }
+  }
+}
+```
+But the majority of users frequently update their git personal access tokens (like once a day), and which means updating the environment resource every time the token is updated.
+
+
+Option 4 : Using Applications.Core/secretStores
+
+Use secretStore to store the username and personal access token and add the secret to the new property recipeConfig in the environment.
+
+```diff
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'dsrp-resources-env-recipes-context-env'
+  location: 'global'
+  properties: {
+    compute: {
+      ...
+    }
+    providers: {
+      ...
+    }
++   recipeConfig: {
++     terraform: {
++       secrets: {
++         "dev.azure.com": {
++            secret: secretStore.id
++          }
++       }
++     }
++   }
+    recipes: {      
+      'Applications.Datastores/mongoDatabases':{
+        default: {
+          templateKind: 'terraform'
+          templatePath: 'https://dev.azure.com/test-private-repo'
+        }
+      }
+    }
+  }
+}
+
+resource secretStore 'Applications.Core/secretStores@2023-10-01-preview' = {
+  name: 'github'
+  properties:{
+-  	app: app.id
+    type: 'generic'
+    data: {
+      'pat': {
+        value: '<personal-access-token>'
+      }
+      'username': {
+        value: '<username>'
+      }
+    }
+  }
+}
+
+```
+But todays secret store implementation is tied to application scope and in this case, secretStore needs to be created before application and environment creation. So we can only use the secretStore resource if we extend it to have global scope(extending to environment scope would create cyclic dependency between environment and secret store).
+
+Option 4 : Using kubernetes secret
```

Comment:
Have we considered leveraging our existing rad credential register API for recipe related credentials?

Reply:
+1

Reply:
UCP credential is designed for IAM credentials. Ideally, UCP credential should not be fetched by RP due to the security reason. For this scenario, using secretstore makes sense.

Reply:
> UCP credential is designed for IAM credentials.

I see. We don't need to use the UCP API if the credentials are only relevant for recipes RP. An extension of the existing command like `rad credential register recipe` instead `rad credential register azure` would provide users the convenience of handing off credentials management without additional mental load of an entirely different command. The backend API for this can be directly implemented in the Recipes RP.

> Ideally, UCP credential should not be fetched by RP due to the security reason. For this scenario, using secretstore makes sense.

Could you share more on the security concern?

> For this scenario, using secretstore makes sense.

Ideally using a secret store resource is easier for Cloud provider credentials (that we manage in UCP) management as well, but it doesn't bring easiness for users since it puts the responsibility of storage on them. So it depends on what we support.